tests/interfaces/test_batch_torch.py

```diff
@@ -0,0 +1,947 @@
+        )
+        assert torch.allclose(res.detach(), expected, atol=tol, rtol=0)
+
+        loss = torch.sum(res)
+        loss.backward()
+        expected = torch.tensor(
+            [
+                -np.sin(x_val) / 2
+                + np.sin(x_val) / 2
+                - np.sin(x_val) * np.cos(y_val) / 2
+                + np.cos(y_val) * np.sin(x_val) / 2,
+                -np.cos(x_val) * np.sin(y_val) / 2 + np.cos(x_val) * np.sin(y_val) / 2,
+            ],
+            dtype=x.grad.dtype,
+            device=torch_device,
+        )
+        assert torch.allclose(x.grad, expected[0], atol=tol, rtol=0)
+        assert torch.allclose(y.grad, expected[1], atol=tol, rtol=0)
+
+    def test_ragged_differentiation(self, torch_device, execute_kwargs, tol):
+        """Tests correct output shape and evaluation for a tape
+        with prob and expval outputs"""
+        if execute_kwargs["gradient_fn"] == "device":
+            pytest.skip("Adjoint differentiation does not yet support probabilities")
+
+        dev = qml.device("default.qubit", wires=2)
+        x_val = 0.543
+        y_val = -0.654
+        x = torch.tensor(x_val, requires_grad=True, device=torch_device)
+        y = torch.tensor(y_val, requires_grad=True, device=torch_device)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0))
+            qml.probs(wires=[1])
+
+        res = execute([tape], dev, **execute_kwargs)[0]
+
+        expected = torch.tensor(
+            [
+                np.cos(x_val),
+                (1 + np.cos(x_val) * np.cos(y_val)) / 2,
+                (1 - np.cos(x_val) * np.cos(y_val)) / 2,
+            ],
+            dtype=res.dtype,
+            device=torch_device,
+        )
+        assert torch.allclose(res.detach(), expected, atol=tol, rtol=0)
+
+        loss = torch.sum(res)
+        loss.backward()
+        expected = torch.tensor(
+            [
+                -np.sin(x_val)
+                + -np.sin(x_val) * np.cos(y_val) / 2
+                + np.cos(y_val) * np.sin(x_val) / 2,
+                -np.cos(x_val) * np.sin(y_val) / 2 + np.cos(x_val) * np.sin(y_val) / 2,
+            ],
+            dtype=x.grad.dtype,
+            device=torch_device,
+        )
+        assert torch.allclose(x.grad, expected[0], atol=tol, rtol=0)
+        assert torch.allclose(y.grad, expected[1], atol=tol, rtol=0)
+
+    def test_sampling(self, torch_device, execute_kwargs):
+        """Test sampling works as expected"""
+        if execute_kwargs["gradient_fn"] == "device" and execute_kwargs["mode"] == "forward":
+            pytest.skip("Adjoint differentiation does not support samples")
+
+        dev = qml.device("default.qubit", wires=2, shots=10)
+
+        with qml.tape.JacobianTape() as tape:
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            qml.sample(qml.PauliZ(0))

```

Comment:
Maybe worth having a test for an error when trying to differentiate a tape with `qml.sample` in it?

Reply:
good idea!

Reply:
done :)